src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -263,6 +264,15 @@ public NodeStateTable<T> ToImmutableAndFree()
                 var hasNonCached = _states.Any(static s => !s.IsCached);
                 return new NodeStateTable<T>(_states.ToImmutableAndFree(), isCompacted: !hasNonCached);
             }
+
+            private (T chosen, EntryState state) GetModifiedItemAndState(T previous, T replacement, IEqualityComparer<T> comparer)
+            {
+                // when comparing an item to check if its modified we explicitly cache the *previous* item in the case where its 
+                // considered to be equal. This ensures that subsequent comparisons are stable across future generation passes.
+                return comparer.Equals(previous, replacement)
```

Comment:
I wouldn't dive into any rabbit holes here but I do wonder if there is a practical way to observe that the user's comparer is not transitive and give a warning. Seems like whatever you could do to check that would introduce undesirable overhead in a release build, though.